runtime/node/src/webapp.ts

```diff
@@ -0,0 +1,192 @@
+
+      onTurnErrorMessage = 'To continue to run this bot, please fix the bot source code.';
+      await turnContext.sendActivity(onTurnErrorMessage, onTurnErrorMessage, InputHints.ExpectingInput);
+
+      // Send a trace activity, which will be displayed in Bot Framework Emulator.
+      await turnContext.sendTraceActivity(
+        'OnTurnError Trace',
+        `${ error }`,
+        'https://www.botframework.com/schemas/error',
+        'TurnError'
+      );
+    } catch (err) {
+      console.error(`\n [onTurnError] Exception caught in sendErrorMessage: ${ err }`);
+    }
+    await conversationState.clear(turnContext);
+    await conversationState.saveChanges(turnContext);
+  };
+  return adapter;
+}
+
+export const getSettings = (projectRoot: string): BotSettings => {
+  // Find settings json file
+  let settings = {} as BotSettings;
+  // load appsettings.json
+  const appsettingsPath = path.join(projectRoot, "settings/appsettings.json");
+  if (fs.existsSync(appsettingsPath)) {
+    const items = require(appsettingsPath);
+    settings = Object.assign(settings, items); // merge settings
+  }
+
+  // load generated settings
+  const generatedPath = path.join(projectRoot, "generated");
+  if (fs.existsSync(generatedPath)) {
+    const generatedFiles = fs.readdirSync(generatedPath);
+    for (let file of generatedFiles) {
+      if (file.endsWith(".json")) {
+        const items = require(path.join(generatedPath, file));
+        settings.luis = Object.assign(settings.luis, items.luis); // merge luis settings
+      }
+    }
+  }
+
+  // load settings from arguments
+  for (let key in argv) {
+    if (key.indexOf(":") >= 0) {
+      const segments: string[] = key.split(":");
+      let base = settings;
+      for (let i = 0; i < segments.length - 1; i++) {
+        const segment = segments[i];
+        if (!base.hasOwnProperty(segment)) {
+          base[segment] = {};
+        }
+        base = base[segment];
+      }
+      base[segments[segments.length - 1]] = argv[key];
+    } else {
+      settings[key] = argv[key];
+    }
+  }
+  settings.MicrosoftAppId = settings.MicrosoftAppId || process.env.MicrosoftAppId;
+  settings.MicrosoftAppPassword = settings.MicrosoftAppPassword || process.env.MicrosoftAppPassword;
+  return settings;
+};
+
+// Create HTTP server.
+const server = restify.createServer();
+const argv = require("minimist")(process.argv.slice(2));
+// prefer the argv port --port=XXXX over process.env because the parent Composer app uses that.
+const port = argv.port || process.env.port || process.env.PORT || 3979;
+
+server.listen(port, (): void => {
+  console.log(
+    `\nGet Bot Framework Emulator: https://aka.ms/botframework-emulator`
+  );
+  console.log(
+    `\nTo talk to your bot, open http://localhost:${ port }/api/messages in the Emulator.`
+  );
+});
+
+const memoryStorage = new MemoryStorage();
+const projectRoot = getProjectRoot();
```

Comment:
As much of this as possible should be moved inside the ComposerBot class so the constructor can have less parameters.